src/async-wrap.cc

```diff
@@ -207,25 +209,23 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
   }
 
   if (ran_init_callback() && !pre_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
     pre_fn->Call(context, 0, nullptr);
     if (try_catch.HasCaught())
       FatalError("node::AsyncWrap::MakeCallback", "pre hook threw");
-    try_catch.SetVerbose(true);
   }
 
   Local<Value> ret = cb->Call(context, argc, argv);
 
-  if (try_catch.HasCaught()) {
-    return Undefined(env()->isolate());
-  }
-
   if (ran_init_callback() && !post_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
+    // TODO(trevnorris): Error reporting in async wrap post callback.
     post_fn->Call(context, 0, nullptr);
     if (try_catch.HasCaught())
       FatalError("node::AsyncWrap::MakeCallback", "post hook threw");
-    try_catch.SetVerbose(true);
+  }
+
+  // If the return value is empty then the callback threw.
+  if (ret.IsEmpty()) {
```

Comment:
Actually, `node::FatalException` is called before the call to `Function::Call` returns in `node::*::MakeCallback`, so when the process' `uncaughtException` handler is called, `node::*::MakeCallback hasn't had the opportunity to call `domain.exit` yet.

So not returning early wouldn't fix the problem, it needs to be fixed in a different way.